zebra/zapi_msg.c

```diff
@@ -1586,8 +1586,45 @@ static void zread_route_add(ZAPI_HANDLER_ARGS)
 	if (CHECK_FLAG(api.message, ZAPI_MESSAGE_SRCPFX))
 		src_p = &api.src_prefix;
 
+
+	/**
+	 * zebra_nhg cached handling to speed up route installs
+	 * for routes being sent with the same nexthops in a row.
+	 */
+	assert(sizeof(api.nexthops) == sizeof(client->inhg_cache.nexthops));
+
+	/* If matches cached nexthops, use that NHE ID now */
+	if (memcmp(api.nexthops, client->inhg_cache.nexthops,
+		   sizeof(client->inhg_cache.nexthops))
+	    == 0) {
+		if (IS_ZEBRA_DEBUG_RECV)
+			zlog_debug("%s: Using cached zebra_nhg", __func__);
+
+		/* Just pass down with ID */
+		nexthop_group_delete(&re->ng);
+		re->nhe_id = client->inhg_cache.nhe_id;
+	} else {
+		/**
+		 * De-ref the old NHE.
+		 * Reset the cached ID to 0 so we know to
+		 * get the next created NHE and use it.
+		 */
+		zebra_nhg_decrement_ref(
+			zebra_nhg_lookup_id(client->inhg_cache.nhe_id));
+		client->inhg_cache.nhe_id = 0;
+	}
+
 	ret = rib_add_multipath(afi, api.safi, &api.prefix, src_p, re);
 
+	if (client->inhg_cache.nhe_id == 0) {
```

Comment:
why defer this second part of the caching operation like this - `ret` isn't being tested, for example. why not just put the creation of a new entry up in the other new block of code?

Reply:
because `rib_add_multipath()` is what actually creates the calls the NHE creation/lookup code. So it has to execute before we can do any caching.

But you're right, the `ret` status needs to be tested here in case we have failed in creation of the `route_entry` in general. If it fails, we shouldn't be caching its result.

Will fix that.